### DISCUSSION TITLE
Can't log in on android with FIDO2 WebAuthn enabled

### DISCUSSION THREAD

**Discussion by jonas-w** [dani-garcia/vaultwarden#4185] [Q&A] [answered]
### Subject of the issue
Logging in on Android fails with a stuck screen on "Validating" and the server logs say:
"2FA token not provided"

### Deployment environment

* vaultwarden version: 1.30.1

* Install method: Docker image

* Clients used: Android 2023.12.0 (both F-Droid and Google Play Store) 

* Reverse proxy and version: nginx 1.22.1 

* MySQL/MariaDB or PostgreSQL version: PostgreSQL 16 

* Other relevant details:

### Expected behaviour

1. Select self-hosted
2. Enter Vaultwarden URL
3. Enter E-Mail Address
4. Enter password
5. Follow FIDO2 WebAuthn instructions
6. login

### Actual behaviour

[...]
5. See: "Validating" and spinner circle, and in the background the FIDO2 WebAuthn instructions are visible but it's loading infinitely.
6. no login

### Troubleshooting data

This is the complete log data from starting the Android app until the "Validating" spinner logo.
```log
vaultwarden  | [2023-12-20 00:57:57.826][request][INFO] GET /api/config/
vaultwarden  | [2023-12-20 00:57:57.826][response][INFO] (config) GET /api/config => 200 OK
vaultwarden  | [2023-12-20 00:57:59.055][request][INFO] GET /api/devices/knowndevice
vaultwarden  | [2023-12-20 00:57:59.058][response][INFO] (get_known_device) GET /api/devices/knowndevice => 200 OK
vaultwarden  | [2023-12-20 00:58:10.488][request][INFO] POST /identity/accounts/prelogin
vaultwarden  | [2023-12-20 00:58:10.489][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
vaultwarden  | [2023-12-20 00:58:13.108][request][INFO] POST /identity/connect/token
vaultwarden  | [2023-12-20 00:58:13.162][error][ERROR] 2FA token not provided
vaultwarden  | [2023-12-20 00:58:13.163][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
```

Maybe relevant: I've tried logging in via the Bitwarden Firefox extension and that worked.

**Comment by BlackDex** [comment]
Check your reverse proxy config if that is configured correctly me
Also check if there are no warnings or errors at the `/admin/diagnostics` page.

Most likely you override some security headers which break the workings of some of the features.

**Reply by jonas-w** [reply]
@BlackDex thanks for your reply.

The /admin/diagnostics does not show any warnings.

The nginx config is basically the same as in your wiki https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples `Nginx - v1.29.0+ (by [@BlackDex](https://github.com/BlackDex))` except my domain name and the correct port set.

**Reply by BlackDex** [reply]
But, does your nginx also have some other settings/mods enabled like ModSecurity or WAF or something else similar?
Those could cause issues like this.

**Reply by jonas-w** [reply]
@BlackDex nothing like that. Only your example file which gets included by nginx.conf.
Here is the top level nginx.conf, the vaultwarden nginx configuration is under conf.d and gets included at the bottom.
<details><summary>Show nginx.conf</summary>

```nginx
user www-data;
worker_processes auto;
pid /run/nginx.pid;
include /etc/nginx/modules-enabled/*.conf;

events {
	worker_connections 1024;
	# multi_accept on;
}
worker_rlimit_nofile 65535;

http {
	sendfile on;
	tcp_nopush on;
	types_hash_max_size 2048;
	server_tokens off;

	# server_names_hash_bucket_size 64;
	# server_name_in_redirect off;
	proxy_headers_hash_max_size 1024;
	proxy_headers_hash_bucket_size 128;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;
    add_header "X-Robots-Tag" "noindex, nofollow" always;

	access_log /var/log/nginx/access.log;
	error_log /var/log/nginx/error.log;

	gzip on;

	upstream php-handler {
	    server unix:/run/php/php8.2-fpm.sock;
	}

include conf.d/*;
}
```
</details>

But as I said earlier, logging in via the web extension works.
I just tested via the Bitwarden Desktop app, and it works there too, but the Bitwarden Desktop app didn't use FIDO2 WebAuthn.

**Comment by stefan0xC** [comment, marked as answer]
What version of Android and what browser are you using? According to the [Bitwarden Help page](https://bitwarden.com/help/setup-two-step-login-fido/) you need to have a FIDO2-supported browser.

**Reply by jonas-w** [reply]
@stefan0xC thank you, that was the problem, my default browser "Firefox Nightly" does not have FIDO2 support, so I needed to temporarily set the default browser to Chrome, and now it worked.

But nonetheless very weird, that it didn't show any error on the client-side.
